wolfcrypt/src/wc_port.c

```diff
@@ -1018,6 +1018,66 @@ size_t wc_strlcat(char *dst, const char *src, size_t dstSize)
 }
 #endif /* USE_WOLF_STRLCAT */
 
+#ifdef SINGLE_THREADED
+int wolfSSL_RefInit(wolfSSL_Ref* ref)
+{
+    ref->count = 1;
+    return 0;
+}
+
+int wolfSSL_RefFree(wolfSSL_Ref* ref)
+{
+    (void)ref;
+    return 0;
+}
+
+int wolfSSL_RefInc(wolfSSL_Ref* ref)
+{
+    ref->count++;
+    return 0;
+}
+
+int wolfSSL_RefDec(wolfSSL_Ref* ref, int* isZero)
+{
+    ref->count--;
+    *isZero = (ref->count == 0);
+    return 0;
+}
+#else
+/* TODO: use atomic operations instead of mutex */
+int wolfSSL_RefInit(wolfSSL_Ref* ref)
+{
+    ref->count = 1;
+    return wc_InitMutex(&ref->mutex);
+}
+
+int wolfSSL_RefFree(wolfSSL_Ref* ref)
+{
+    return wc_FreeMutex(&ref->mutex);
+}
+
+int wolfSSL_RefInc(wolfSSL_Ref* ref)
+{
+    int ret = wc_LockMutex(&ref->mutex);
+    if (ret == 0) {
+        ref->count++;
+        wc_UnLockMutex(&ref->mutex);
+    }
+    return ret;
+}
+
+int wolfSSL_RefDec(wolfSSL_Ref* ref, int* isZero)
+{
+    int ret = wc_LockMutex(&ref->mutex);
+    if (ret == 0) {
+        ref->count--;
+        *isZero = (ref->count == 0);
+        wc_UnLockMutex(&ref->mutex);
+    }
+    return ret;
+}
```

Comment:
Would be good to log erroneous locking here instead of in functions higher in the stack.

Reply:
WOLFSSL_MSG calls added.